### PR TITLE
feat: Arrays.sort for struct arrays with SortKey builder

### LINKED ISSUE #56: brainstorm: Arrays.sort for struct arrays with SortKey builder
# Brainstorm: Arrays.sort for Struct Arrays with SortKey

**Date:** 2026-03-20
**Status:** Ready for planning
**Related:** Issue #55

## What We're Building

A `SortKey` builder type and new `Arrays.sort` / `Arrays.reverse` overloads that let JCP users sort struct arrays by one or more fields, with per-field direction control and nested field path support.

## Why

`Arrays.sort(Object[])` requires elements to implement `Comparable`. `StructData` does not, so users cannot sort struct arrays at all today. A field-name-based API solves this without requiring lambda support.

### SortKey builder

```java
SortKey.by("name")               // ascending by default
SortKey.by("name").asc()
SortKey.by("name").desc()
SortKey.by("address.city").desc() // nested field path via dot notation
```

### Arrays.sort overloads

```java
// Single field
Arrays.sort(persons, SortKey.by("age"))

// Multiple fields (primary, secondary, ...)
Arrays.sort(persons, SortKey.by("lastName"), SortKey.by("firstName"))

// Mixed directions
Arrays.sort(persons, SortKey.by("age").desc(), SortKey.by("name").asc())
```

### Arrays.reverse

```java
Arrays.reverse(a)  // reverses any array in-place, returns the array
```

### Deliverables
- `SortKey` class in `module/base/arrays/` with `by(String fieldPath)`, `.asc()`, `.desc()`
- `Arrays.sort(Object[], SortKey...)` overload — struct arrays only
- `Arrays.reverse(int[])`, `Arrays.reverse(String[])`, `Arrays.reverse(boolean[])`, `Arrays.reverse(double[])`, `Arrays.reverse(Object[])` overloads
- `BaseModuleRegister` updated to expose new overloads
- Tests for eval mode covering: single field, multi-field, nested path, ASC/DESC, null-field handling, reverse

### Out of Scope
- Sorting non-struct `Object[]` with `SortKey` (e.g. boxed `Integer[]`)
- Lambda/function-reference comparator (tracked separately, blocked on lambda support)
- Compile mode support for `SortKey` (can be added once eval is stable)

## Key Decisions

| Decision | Choice | Rationale |
|----------|--------|-----------|
| Field path separator | Dot notation (`"address.city"`) | Familiar, mirrors JCP field access syntax |
| Applies to | Struct arrays only | Keeps scope focused; primitive arrays have dedicated overloads |
| Missing/null field | Nulls sort last (both ASC and DESC) | Predictable, non-crashing behavior |
| Direction default | ASC | Industry standard default |
| Reverse | Dedicated `Arrays.reverse()` | Orthogonal to sorting; useful independently |
| SortKey location | `module/base/arrays/SortKey.java` | Co-located with `Arrays.java` |

## Field Path Resolution

For a path like `"address.city"`:
1. Split on `.`
2. Walk `StructData.getField()` for each segment
3. If any segment returns null or a non-`StructData` mid-path, treat as null (sort last)
4. Final value compared via natural ordering of the primitive `Data` value

Supported leaf types for comparison: `int`, `String`, `boolean`, `double` (existing `SystemDataType` primitives).

## Open Questions

1. Should `SortKey` be exposed as a first-class JCP `DataType` (so it appears in the type system), or remain a pure Java helper invoked via static method call?
2. Should `Arrays.reverse` return the array (consistent with existing `sort` overloads) or `void`?

## Next Steps

Run `/gw-plan` to create implementation plan.

---

## Enhanced Implementation Plan

**Enhanced plan document:** `docs/plans/2026-03-20-feat-arrays-sort-struct-sortkey-plan-enhanced.md`
**Original plan:** `docs/plans/2026-03-20-feat-arrays-sort-struct-sortkey-plan.md`
**Review:** `docs/plans/2026-03-20-feat-arrays-sort-struct-sortkey-plan-review.md`

---
title: "feat: Arrays.sort for struct arrays with SortKey builder (Enhanced)"
type: feat
date: 2026-03-20
issue: 56
review_date: 2026-03-20
original_plan: docs/plans/2026-03-20-feat-arrays-sort-struct-sortkey-plan.md
---

# feat: Arrays.sort for Struct Arrays with SortKey Builder (Enhanced)

> **Note:** Enhanced version incorporating feedback from plan review on 2026-03-20.
> Primary fix: replaced `SortKey...` varargs with fixed-arity overloads to match JCP's
> reflection-based dispatch mechanism.

## Overview

Add a `SortKey` builder type and new `Arrays.sort` / `Arrays.reverse` overloads to the JCP base
module. This lets users sort struct arrays by one or more fields with per-field direction
(ASC/DESC), dot-notation nested field paths, and null-last semantics — without requiring lambda
support.

### SortKey — pure Java builder, not a JCP DataType

`SortKey` is a plain Java class registered alongside `Arrays` in `BaseModuleRegister`. It does
**not** need to be a `SystemDataType` entry. `ClassConverter` will auto-register it as a DataType
named `"SortKey"` (via `DataTypeUtils.getDataTypeAndCreateOnMissing`) when it appears as a
parameter type. No manual type-system changes are required.

```
module/base/arrays/
  Arrays.java          ← add sort(Object[], SortKey) + sort(Object[], SortKey, SortKey) +
                          sort(Object[], SortKey, SortKey, SortKey) + reverse overloads
  SortKey.java         ← new: builder class
```

### Why fixed-arity overloads instead of varargs

JCP's `ClassConverter.getParameterDefs()` introspects methods via `method.getParameters()`.
Java varargs `SortKey...` is seen as `SortKey[]` — a single fixed-arity array parameter. The
runtime evaluator `FunCallEvaluator.hasSameParameterDefinition()` matches by
`parameterDefs.length == arguments.length`. A caller passing 2 `SortKey` values has
`argumentData.length == 2`, but the registered function has `parameterDefs.length == 1`.
This causes `UndeclaredException` at runtime.

**Solution:** Follow the existing `Arrays.java` pattern of explicit typed overloads:

```java
public static Object[] sort(Object[] a, SortKey k1)
public static Object[] sort(Object[] a, SortKey k1, SortKey k2)
public static Object[] sort(Object[] a, SortKey k1, SortKey k2, SortKey k3)
```

Three overloads cover all practical cases (primary, primary+secondary, three-level).
The internal implementation delegates to a shared private method that accepts `SortKey[]`.

### Field path resolution

For a path `"address.city"`:
1. Split on `"."` → `["address", "city"]`
2. Starting from the `StructData` element, call `getField(segment)` for each segment
3. If any intermediate result is `null` or not a `StructData`, return `null` (→ sorts last)
4. The final `Data` value's `.get()` is compared via its natural `Comparable` order

Supported leaf types: `Integer` (INT), `String`, `Boolean`, `Double` — all box types implement
`Comparable`. Unsupported leaf types treat the value as `null` and sort last.

### Comparator construction (internal)

The private shared method composes per-field comparators:

```java
private static Object[] sortByKeys(Object[] a, SortKey... keys) {
    Comparator<Object> comparator = (x, y) -> 0;
    for (SortKey key : keys) {
        comparator = comparator.thenComparing(key.toComparator());
    }
    java.util.Arrays.sort(a, comparator);
    return a;
}
```

Each `SortKey.toComparator()` resolves the field path, applies null-last logic, then reverses if
`DESC`.

### Arrays.reverse

Five typed overloads (int[], String[], boolean[], double[], Object[]) each reverse in-place and
return the array — consistent with existing `sort` overloads.

### BaseModuleRegister

Add `SortKey.class` to the registration list. `ClassConverter` will process it and register
`SortKey` as a DataType so it can appear as a parameter type in `Arrays.sort` overloads.

### Phase 1: SortKey builder

**Deliverables:**
- [ ] Create `SortKey.java` in `module/base/arrays/`
  - `public static SortKey by(String fieldPath)` — factory method, ASC default
  - `.asc()` / `.desc()` — fluent direction setters, return `this`
  - `Comparator<Object> toComparator()` — package-private, used by `Arrays`
  - Private field path resolution: split on `"."`, walk `StructData.getField()`, null-last
- [ ] Unit-test `SortKey.toComparator()` directly (no JCP runtime needed):
  - Single-level field, int leaf, ASC and DESC
  - Single-level field, String leaf
  - Nested path `"address.city"`
  - Missing field → sorts last (both ASC and DESC)
  - Non-StructData element → sorts last
  - Empty field path edge case

### Phase 2: Arrays.sort with SortKey (fixed-arity overloads)

**Deliverables:**
- [ ] Add to `Arrays.java`:
  - `public static Object[] sort(Object[] a, SortKey k1)`
  - `public static Object[] sort(Object[] a, SortKey k1, SortKey k2)`
  - `public static Object[] sort(Object[] a, SortKey k1, SortKey k2, SortKey k3)`
  - Private `sortByKeys(Object[] a, SortKey... keys)` — shared implementation
- [ ] Register `SortKey.class` in `BaseModuleRegister.classToRegister()`
- [ ] Integration tests via `FunctionCallExpression` dispatch (not direct Java calls):
  - 1-key sort (validates `sort(Object[], SortKey)` overload resolves)
  - 2-key sort (validates `sort(Object[], SortKey, SortKey)` overload resolves)
  - Sort with nested path via dispatch
  - Null-field element via dispatch

### Phase 3: Arrays.reverse

**Deliverables:**
- [ ] Add `reverse(int[])`, `reverse(String[])`, `reverse(boolean[])`, `reverse(double[])`,
  `reverse(Object[])` to `Arrays.java` — each reverses in-place, returns the array
- [ ] Tests for all five overloads:
  - Empty array
  - Single-element array
  - Even-length array
  - Odd-length array

### Phase 4: Coverage verification

- [ ] Run `mvn verify -pl core` — confirm JaCoCo ≥ 80% instruction and branch
- [ ] Fix any coverage gaps (target: SortKey path resolution branches, all `sort` overloads)

### Functional

- [ ] `Arrays.sort(persons, SortKey.by("age"))` sorts struct array by int field ascending
- [ ] `Arrays.sort(persons, SortKey.by("name").desc())` sorts descending
- [ ] `Arrays.sort(persons, SortKey.by("lastName"), SortKey.by("firstName"))` sorts by two fields
- [ ] `Arrays.sort(persons, SortKey.by("lastName"), SortKey.by("firstName"), SortKey.by("age"))` sorts by three fields
- [ ] `Arrays.sort(persons, SortKey.by("address.city"))` resolves nested field path
- [ ] Elements with a missing/null field sort to the end (null-last, both ASC and DESC)
- [ ] `Arrays.reverse(intArray)` reverses in-place and returns the array
- [ ] `Arrays.reverse(objectArray)` works for struct arrays
- [ ] Existing `Arrays.sort(int[])` / `sort(String[])` etc. are unaffected
- [ ] All three `sort(Object[], SortKey...)` overloads are callable via `FunctionCallExpression` dispatch without `UndeclaredException`

### Non-Functional

- [ ] No new `SystemDataType` entries required
- [ ] `SortKey` is not exposed as a JCP DSL keyword — it is a library type only
- [ ] JaCoCo thresholds pass (≥ 80% instruction, ≥ 80% branch)
- [ ] No compile-mode changes (out of scope)

## Dependencies

- No blocked dependencies — self-contained within `core/`
- `StructData.getField(String)` already exists and is used as-is
- `ClassConverter.registerClass()` auto-registers `SortKey` as a DataType when encountered as a parameter type — no manual registration beyond adding to `BaseModuleRegister`

## Risk Analysis

| Risk | Likelihood | Impact | Mitigation |
|------|------------|--------|------------|
| `SortKey` DataType not auto-registered before `Arrays` methods are scanned | M | H | Register `SortKey.class` **before** `Arrays.class` in `BaseModuleRegister.classToRegister()` list |
| Fixed-arity overloads cause `FunctionAmbiguityException` when 2-key and 3-key overloads both match | L | H | Java dispatch passes exact argument count to `getParameterDefs`; `hasSameParameterDefinition` checks exact length — no ambiguity possible across different arities |
| `ClassConverter` unwraps `Data.get()` before passing to Java method; `SortKey` is the raw value | L | M | `AbstractModuleFunction.doFunction` calls `getArgumentValues` which calls `.get()` on each `Data`; since `SortKey` is stored as `AnyData<SortKey>`, `.get()` returns the `SortKey` object correctly |
| Nested path on non-struct mid-node silently wrong | L | M | Covered by null-last rule; explicit test: path through a non-struct field |
| Coverage drops below 80% threshold | L | H | Tests cover: all sort overloads, all reverse overloads, all SortKey direction/path branches, null-field case, unsupported leaf type |

## Testing Strategy

**Unit tests** (no JCP runtime — test Java logic directly):
- `SortKeyTest`: path resolution, direction, null-last, nested path, unsupported leaf
- `ArraysReverseTest`: all five typed overloads, edge cases

**Integration tests** (via `FunctionCallExpression` — validates full dispatch chain):
- `ArraysSortStructTest`: all three sort overloads, nested path, null-field, multi-key ordering
- Pattern: follow `ArraysTest.evalOnArray()` helper — seed array in context, call via `FunctionCallExpression`, assert result

## Changes from Original Plan

1. **Replaced `SortKey...` varargs with fixed-arity overloads** — `sort(Object[], SortKey)`, `sort(Object[], SortKey, SortKey)`, `sort(Object[], SortKey, SortKey, SortKey)`. Root cause: JCP's reflection-based dispatch (`ClassConverter` + `FunCallEvaluator`) does not handle varargs; callers with N `SortKey` args would cause `UndeclaredException`.
2. **Corrected risk mitigation for varargs** — Original mitigation cited Java compile-time overload resolution, which is irrelevant to JCP runtime dispatch. Updated to document the fixed-arity approach as the actual solution.
3. **Added dispatch integration tests as an explicit acceptance criterion** — Direct Java unit tests of `SortKey.toComparator()` do not catch runtime dispatch failures. Integration tests via `FunctionCallExpression` are required.
4. **Added SortKey registration ordering note** — `SortKey.class` must precede `Arrays.class` in the `BaseModuleRegister` list to ensure the `SortKey` DataType exists before `Arrays` methods are scanned.
5. **Expanded testing strategy** — Separated unit tests (pure Java) from integration tests (JCP dispatch), with explicit patterns for each.

## References

- Original plan: `docs/plans/2026-03-20-feat-arrays-sort-struct-sortkey-plan.md`
- Plan review: `docs/plans/2026-03-20-feat-arrays-sort-struct-sortkey-plan-review.md`
- Issue: https://github.com/elminsterjimmy/jcp/issues/56

### PR DESCRIPTION
**Pull request by elminsterjimmy**
Closes #56

## Summary

- `SortKey.by("field").asc()/.desc()` — fluent builder for struct sort keys with dot-notation nested field paths
- `Arrays.sort(Object[], SortKey)`, `sort(Object[], SortKey, SortKey)`, `sort(Object[], SortKey, SortKey, SortKey)` — sort struct arrays by 1–3 fields
- `Arrays.reverse(int[]/String[]/boolean[]/double[]/Object[])` — in-place reverse for all array types
- Null-last semantics: missing or non-navigable fields sort to the end in both ASC and DESC

## Key Design Decisions

**Fixed-arity overloads instead of varargs:** JCP's `ClassConverter` sees `SortKey...` as `SortKey[]` (1 parameter). `FunCallEvaluator` matches by exact arity — a caller passing 2 `SortKey` values would get `UndeclaredException`. Three fixed-arity overloads follow the existing `Arrays.java` pattern.

**SortKey registered before Arrays:** `SortKey.class` precedes `Arrays.class` in `BaseModuleRegister` so its DataType is available when `Arrays` method signatures are scanned.

## Changes

- `SortKey.java` — new builder class with field path resolution
- `Arrays.java` — 3 sort overloads + 5 reverse overloads
- `BaseModuleRegister.java` — add `SortKey.class` before `Arrays.class`
- `SortKeyTest.java` — 11 unit tests (no runtime needed): ASC/DESC, nested path, null-last, non-struct element
- `ArraysSortStructTest.java` — 5 integration tests via `FunctionCallExpression` dispatch
- `ArraysTest.java` — 9 reverse tests added

## Testing

- 42 new tests added; 1170 total pass
- JaCoCo coverage checks pass (≥ 80% instruction and branch)

### CODE REVIEW

File: core/src/main/java/com/elminster/jcp/module/base/arrays/SortKey.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        this.descending = false;
+    }
+
+    /** Creates a sort key for the given field path, ascending by default. */
+    public static SortKey by(String fieldPath) {
+        return new SortKey(fieldPath);
+    }
+
+    /** Sets direction to ascending. Returns {@code this} for chaining. */
+    public SortKey asc() {
```
Comment (elminsterjimmy) [inline core/src/main/java/com/elminster/jcp/module/base/arrays/SortKey.java:57]: **Mutability concern:** `asc()` and `desc()` mutate `this` and return it. If a caller reuses the same `SortKey` instance, a later `.desc()` call changes the sort direction of any previously created comparator that still holds a reference to the key (since `toComparator()` captures `this.descending` via a closure over `this`, not a snapshot). For the current use case — single-call creation like `SortKey.by("age").desc()` — this is safe. But it's worth documenting that `SortKey` instances should not be reused after calling `asc()`/`desc()`. Consider either (a) adding a note to the Javadoc, or (b) making `asc()`/`desc()` return a new `SortKey` (true builder pattern).

File: core/src/main/java/com/elminster/jcp/module/base/arrays/SortKey.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+     * Returns {@code null} if the element is not a {@link StructData}, any path segment is
+     * missing, any intermediate node is not a {@link StructData}, or the leaf value is not
+     * {@link Comparable}.
+     */
+    @SuppressWarnings("unchecked")
+    private Comparable<Object> resolveField(Object element) {
+        if (!(element instanceof StructData)) {
+            return null;
+        }
+        String[] segments = fieldPath.split("\\.");
```
Comment (elminsterjimmy) [inline core/src/main/java/com/elminster/jcp/module/base/arrays/SortKey.java:100]: **Edge case — empty `fieldPath`:** `SortKey.by("")` will call `"".split("\\.")` which returns `[""]`. The loop doesn't execute (length-1 == 0), and then `current.getField("")` is called. `StructData.getField()` returns `null` for `""`, so this resolves to `null` (sorts last) — safe but silently wrong. A check like `if (fieldPath == null || fieldPath.isEmpty()) return null;` or throwing `IllegalArgumentException` in `by()` would make the API contract explicit.

File: core/src/main/java/com/elminster/jcp/module/base/arrays/Arrays.java [hunk 3]
```diff
@@ -138,11 +141,83 @@ public static double[] sort(double[] a) {
...
+    /**
+     * Sorts a struct array in-place by three fields and returns it.
+     * Elements whose field is missing or non-navigable sort last.
+     */
+    public static Object[] sort(Object[] a, SortKey k1, SortKey k2, SortKey k3) {
+        return sortByKeys(a, k1, k2, k3);
+    }
+
+    private static Object[] sortByKeys(Object[] a, SortKey... keys) {
+        java.util.Comparator<Object> comparator = (x, y) -> 0;
```
Comment (elminsterjimmy) [inline core/src/main/java/com/elminster/jcp/module/base/arrays/Arrays.java]: **Minor: identity comparator is unnecessary.** Starting with `(x, y) -> 0` and then calling `.thenComparing()` on it is correct but adds a vacuous first comparison. A cleaner approach builds the comparator from the first key directly:
```java
Comparator<Object> comparator = keys[0].toComparator();
for (int i = 1; i < keys.length; i++) {
    comparator = comparator.thenComparing(keys[i].toComparator());
}
```
This avoids an extra comparator call per element. Not a bug — just a minor readability and micro-efficiency improvement.

File: core/src/main/java/com/elminster/jcp/module/base/arrays/Arrays.java [hunk 4]
```diff
@@ -138,11 +141,83 @@ public static double[] sort(double[] a) {
...
+    }
+
+    // --- reverse ---
+
+    /**
+     * Reverses the array in-place and returns it.
+     */
+    public static int[] reverse(int[] a) {
+        for (int i = 0, j = a.length - 1; i < j; i++, j--) {
+            int tmp = a[i]; a[i] = a[j]; a[j] = tmp;
```
Comment (elminsterjimmy) [inline core/src/main/java/com/elminster/jcp/module/base/arrays/Arrays.java]: **Style: multiple statements on one line.** The pattern `int tmp = a[i]; a[i] = a[j]; a[j] = tmp;` puts three statements on one line. The rest of the codebase uses one-statement-per-line style. Consider splitting:
```java
int tmp = a[i];
a[i] = a[j];
a[j] = tmp;
```
This applies to all five `reverse` overloads.

**Comment by elminsterjimmy** [comment]
## Code Review Summary

**Files reviewed:** 10 | **Lines changed:** +1107 -2

---

### ✅ Strengths

- **Correct dispatch design:** Fixed-arity overloads correctly solve the varargs incompatibility with JCP's reflection-based dispatch — this was the key insight from the plan review and it's implemented exactly right
- **Field path resolution:** The walk-and-null-out algorithm is clean, handles all edge cases (non-struct element, missing field, non-struct mid-node, non-Comparable leaf), and is well-documented
- **Registration ordering:** `SortKey.class` registered before `Arrays.class` in `BaseModuleRegister` — correctly prevents the DataType-not-found failure at scan time
- **Two-tier test strategy:** Clean separation between unit tests (pure Java, no runtime) and integration tests (full `FunctionCallExpression` dispatch chain) — both layers are needed and present
- **`@SuppressWarnings("unchecked")` scoped minimally** to the private method that needs it
- **All 1170 tests pass, JaCoCo thresholds met**

---

### ⚠️ Issues Found

**1. `SortKey` mutability (SortKey.java:43)**
`asc()`/`desc()` mutate `this`. The `toComparator()` lambda captures `this.descending` by reference (via the enclosing `SortKey` instance), not by snapshot. So if a caller reuses a key:
```java
SortKey k = SortKey.by("age");
sort(arr1, k.desc());  // k is now desc
sort(arr2, k);         // still desc — surprising
```
**Recommendation:** Either document the single-use expectation, or make `asc()`/`desc()` return a new `SortKey`.

**2. Empty `fieldPath` silently sorts-last (SortKey.java:83)**
`SortKey.by("")` calls `"".split("\\.")` → `[""]` → `getField("")` → `null` → sorts last. Safe but silently wrong. **Recommendation:** Add `Objects.requireNonNull` or `IllegalArgumentException` in `by()` for null/empty paths.

---

### 💡 Suggestions (non-blocking)

**3. Identity comparator in `sortByKeys` (Arrays.java:176)**
`(x, y) -> 0` as the seed is correct but adds a vacuous comparison per element. Starting from `keys[0].toComparator()` and looping from index 1 is equivalent and slightly cleaner.

**4. Multi-statement lines in `reverse` (Arrays.java:191)**
`int tmp = a[i]; a[i] = a[j]; a[j] = tmp;` — three statements on one line. The rest of the codebase uses one-statement-per-line. Consider splitting across lines for consistency.

---

### 📋 Checklist

- [x] Code follows project conventions
- [x] Tests added (unit + integration)
- [x] JaCoCo coverage maintained
- [x] No security issues
- [x] No debug statements
- [x] No TODO/FIXME
- [ ] `SortKey` mutability documented or fixed (issue #1)
- [ ] Empty fieldPath guard (issue #2, optional)

**Inline comments posted:** 4

Overall: the implementation is solid and production-ready. Issues #1 and #2 are worth addressing before merge.

File: core/src/main/java/com/elminster/jcp/module/base/arrays/SortKey.java (hunk 1 above)
Comment (elminsterjimmy) [reply core/src/main/java/com/elminster/jcp/module/base/arrays/SortKey.java:57]: Fixed: `asc()` and `desc()` now return a new `SortKey` instance. `descending` is now `final`.

File: core/src/main/java/com/elminster/jcp/module/base/arrays/SortKey.java (hunk 2 above)
Comment (elminsterjimmy) [reply core/src/main/java/com/elminster/jcp/module/base/arrays/SortKey.java:100]: Fixed: `by()` now throws `IllegalArgumentException` for null or empty `fieldPath`.

File: core/src/main/java/com/elminster/jcp/module/base/arrays/Arrays.java (hunk 3 above)
Comment (elminsterjimmy) [reply core/src/main/java/com/elminster/jcp/module/base/arrays/Arrays.java]: Fixed: `sortByKeys` now starts from `keys[0].toComparator()` and loops from index 1.

File: core/src/main/java/com/elminster/jcp/module/base/arrays/Arrays.java (hunk 4 above)
Comment (elminsterjimmy) [reply core/src/main/java/com/elminster/jcp/module/base/arrays/Arrays.java]: Fixed: all five `reverse()` overloads now use one statement per line for the swap.

**Comment by elminsterjimmy** [comment]
Addressed all review feedback:

- **`SortKey` mutability:** `asc()`/`desc()` now return new instances; `descending` is `final`
- **Empty `fieldPath` guard:** `by()` throws `IllegalArgumentException` for null/empty input
- **Identity comparator:** `sortByKeys` now builds from `keys[0].toComparator()` directly
- **Reverse swap style:** All five `reverse()` overloads split to one statement per line

1170 tests pass, JaCoCo thresholds met. Commit: 5131b71

File: core/src/main/java/com/elminster/jcp/module/base/arrays/SortKey.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+ *   SortKey.by("address.city").asc()
+ * </pre>
+ *
+ * @author jgu
+ * @version 1.0
+ */
+public class SortKey {
+
+    private final String fieldPath;
+    private final boolean descending;
```
Comment (elminsterjimmy) [inline core/src/main/java/com/elminster/jcp/module/base/arrays/SortKey.java]: it could be better to have the direction instead of the boolean descending.

File: core/src/main/java/com/elminster/jcp/module/base/arrays/SortKey.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        if (!(element instanceof StructData)) {
+            return null;
+        }
+        String[] segments = fieldPath.split("\\.");
+        StructData current = (StructData) element;
+
+        for (int i = 0; i < segments.length - 1; i++) {
+            Data field = current.getField(segments[i]);
+            if (field == null || !(field.get() instanceof StructData)) {
+                return null;
```
Comment (elminsterjimmy) [inline core/src/main/java/com/elminster/jcp/module/base/arrays/SortKey.java]: it should be raise an error instead of retuning null cause it is an illegal field path

**Comment by elminsterjimmy** [comment]
Addressed both review comments:

- **Direction enum:** Replaced `boolean descending` with `SortKey.Direction` enum (`ASC`/`DESC`). Internal field, constructor, and `toComparator()` updated; `asc()`/`desc()` still return new immutable instances.
- **Illegal field path:** `resolveField` now throws `IllegalArgumentException` when an intermediate path segment is missing or non-navigable (e.g., navigating `x.sub` when `x` is a String, or referencing an unknown segment). Non-`StructData` elements and missing/non-Comparable leaves still sort last — only structurally illegal paths throw.
- Added tests covering both behaviors; 1174 tests pass, JaCoCo thresholds met.

Commit: 5f0ad43
